File: docs/test-cases/model-dos.py
# Test case: model-dos (LLM04:2025)
import anthropic
from flask import Flask, request, jsonify

app = Flask(__name__)
client = anthropic.Anthropic()

conversation_history = {}

@app.route("/chat", methods=["POST"])
def chat():
    user_id = request.json.get("user_id")
    message = request.json.get("message")  # BUG: no length limit on input

    # BUG: conversation history grows unbounded — no context trimming
    if user_id not in conversation_history:
        conversation_history[user_id] = []
    conversation_history[user_id].append({"role": "user", "content": message})

    # BUG: no max_tokens, no rate limiting, no timeout
    response = client.messages.create(
        model="claude-opus-4-6",
        messages=conversation_history[user_id]
    )

    reply = response.content[0].text
    conversation_history[user_id].append({"role": "assistant", "content": reply})
    return jsonify({"reply": reply})

File: docs/test-cases/model-dos.rs
// Test case: model-dos (LLM04:2025)
use axum::{routing::post, Json, Router};
use anthropic_sdk::{Client, MessagesRequestBuilder};
use serde::{Deserialize, Serialize};

#[derive(Deserialize)]
struct ChatRequest {
    user_id: String,
    message: String,
}

#[derive(Serialize)]
struct ChatResponse {
    reply: String,
}

async fn chat(Json(req): Json<ChatRequest>) -> Json<ChatResponse> {
    let client = Client::from_env().unwrap();

    // BUG: req.message has no length cap — axum's default body limit is
    // the only guard, and nothing trims the prompt before forwarding
    let _user = req.user_id;

    // BUG: no per-IP or per-key rate limit (e.g. tower_governor); a single
    // caller can hammer /chat and drain the Anthropic spend budget
    loop {
        let result = MessagesRequestBuilder::default()
            .model("claude-opus-4-5".to_string())
            // BUG: max_tokens omitted — responses run to the model ceiling,
            // and on any transient error we retry forever with no backoff
            .messages(vec![serde_json::json!({
                "role": "user",
                "content": req.message,
            })])
            .build()
            .unwrap()
            .send(&client)
            .await;

        match result {
            Ok(resp) => return Json(ChatResponse { reply: resp.content }),
            Err(_) => continue, // retry forever
        }
    }
}

#[tokio::main]
async fn main() {
    let app = Router::new().route("/chat", post(chat));
    let listener = tokio::net::TcpListener::bind("0.0.0.0:8080").await.unwrap();
    axum::serve(listener, app).await.unwrap();
}
